nexus-sdk/src/main/java/io/nexusrpc/handler/ServiceHandler.java

```diff
@@ -0,0 +1,192 @@
+    if (handler == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    Object result = handler.fetchResult(context, details);
+    return resultToContent(result);
+  }
+
+  private HandlerResultContent resultToContent(Object result) {
+    try {
+      Serializer.Content output = serializer.serialize(result);
+      HandlerResultContent.Builder contentBuilder = HandlerResultContent.newBuilder();
+      contentBuilder.setData(output.getData());
+      contentBuilder.getHeaders().putAll(output.getHeaders());
+      return contentBuilder.build();
+    } catch (Exception e) {
+      throw new RuntimeException("Failed serializing result", e);
+    }
+  }
+
+  @Override
+  public OperationInfo fetchOperationInfo(
+      OperationContext context, OperationFetchInfoDetails details)
+      throws UnrecognizedOperationException, OperationNotFoundException {
+    ServiceImplInstance instance = instances.get(context.getService());
+    if (instance == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    OperationHandler<Object, Object> handler =
+        instance.getOperationHandlers().get(context.getOperation());
+    if (handler == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    return handler.fetchInfo(context, details);
+  }
+
+  @Override
+  public void cancelOperation(OperationContext context, OperationCancelDetails details)
+      throws UnrecognizedOperationException, OperationNotFoundException {
+    ServiceImplInstance instance = instances.get(context.getService());
+    if (instance == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    OperationHandler<Object, Object> handler =
+        instance.getOperationHandlers().get(context.getOperation());
+    if (handler == null) {
+      throw new UnrecognizedOperationException(context.getService(), context.getOperation());
+    }
+    handler.cancel(context, details);
+  }
+
+  /** Builder for operation start details. */
+  public static class Builder {
+    private final List<ServiceImplInstance> instances;
+    private @Nullable Serializer serializer;
+
+    private Builder() {
+      this.instances = new ArrayList<>();
+    }
+
+    private Builder(ServiceHandler handler) {
+      // Order does not matter for instances except in case of validation
+      // errors.
+      instances = new ArrayList<>(handler.instances.values());
+      serializer = handler.serializer;
+    }
+
+    /** Get instances to mutate. */
+    public List<ServiceImplInstance> getInstances() {
+      return instances;
+    }
+
+    /** Add a service instance. */
+    public Builder addInstance(ServiceImplInstance instance) {
+      instances.add(instance);
+      return this;
+    }
+
+    /** Serializer. Required. */
+    public Builder setSerializer(Serializer serializer) {
```

Comment:
There should be a default serializer for the server IMHO and we shouldn't require setting this per service.
How is this going to integrate with the Temporal SDK? We should be using data converters in the Temporal context.

Reply:
> There should be a default serializer for the server IMHO and we shouldn't require setting this per service.

You can't have a default serializer in Java IMO because there there is no JSON support in the standard library. That's why it's required. We should probably offer a Jackson and maybe a Gson one as a separate project/JAR.

> How is this going to integrate with the Temporal SDK? We should be using data converters in the Temporal context.

Yes, the SDK should create a serializer that works with data converter and set this when building the service handler. I don't think users of Temporal SDK workers are going to build service handlers, rather Temporal will build one service handler based on the service impls registered. It is at that time it can set the data-converter-based serializer.

Reply:
I don't even want us to have a default serializer at this point, it's fine to wait with that.
I'm just saying that the service handler would be backed by a server / worker and can inherit the serializer from whatever it's embedded in.

If this abstraction isn't exposed to the Temporal SDK, this works for me.

Reply:
Yes the SDK won't expose the `Serializer` when registering services on a worker